docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,304 @@
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU corev1.ResourceName = "cpu"
+    AutoscalerResourceMemory corev1.ResourceName = "memory"
+    AutoscalerResourceGPU corev1.ResourceName = "gpu"
+)
+
+// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineStatus struct {
+    Capacity corev1.ResourceList `json:"capacity,omitempty"`
+}
+
+// AutoscalerMachineTemplateStatus defines information that is used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineTemplateStatus struct {
+    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
+}
+
+// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
+type DockerMachineTemplateStatus struct {
+    Autoscaler AutoscalerMachineTemplateStatus `json:"autoscaler,omitempty"`
+}
+
+// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
+type DockerMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
+    Status DockerMachineTemplateStatus `json:"status,omitempty"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/api/core/v1`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  template:
+    spec: {}
+status:
+  autoscaler:
+    machine:
+      capacity:
+        memory: 500mb
+        cpu: "1"
+        gpu: "1"
+```
+
+#### MachineSet and MachineDeployment Annotations
+
+In cases where a user needs to provide specific resource information for a
+MachineSet or MachineDeployment, or in cases where an infrastructure provider
+has not yet added the Infrastructure Machine Template status changes, they
+may use annotations to provide the information. The annotation values match
+API that is used in Infrastructure Machine Template, e.g. the memory and cpu
+annotations allow a `resource.Quantity` value and the gpu annotation allows the
+count of GPUs per instance.
+
+If a user wishes to specify the resource capacity through annotations, they
+may do by adding the following to any MachineSet or MachineDeployment (it is not required on both)
+that are participating in autoscaling:
+
+```
+kind: <MachineSet or MachineDeployment>
+metadata:
+  annotations:
+      autoscaler.cluster.x-k8s.io/machine-capacity-gpu: "1"
+      autoscaler.cluster.x-k8s.io/machine-capacity-memory: "500mb"
+      autoscaler.cluster.x-k8s.io/machine-capacity-cpu: "1"
```

Comment:
```suggestion
      capacity.cluster.x-k8s.io/cpu: "1"
```
Maybe?

Reply:
i'd be ok with this, @sbueringer and @fabriziopandini inspired the original change, i'd like to get their opinion as well.

Reply:
I'm not sure what we had before the current one. It would be fine for me to shorten it.

Reply:
I think we should keep the word `autoscaler` in the annotation as this annotation is on the MD/MS object, so without that word the purpose of these annotations might not become clear. So either what it is right now, or `autoscaler.cluster.x-k8s.io/cpu: "1"` if we want to shorten it

Reply:
I think the reason for dropping autoscaler is actually to make the annotation also usable outside of the autoscaler use case (especially for the struct in status). Otherwise it would always be a field which is only there for the autoscaler.

Reply:
oh ok I thought it was meant to be only for autoscaler. 
> especially for the struct in status

I didn't get this part?

Reply:
I just meant the same applies to the `status.autoscaler.machine`. You would never use those fields for other use cases then autoscaler. 
Note: I'm just guessing about @vincepri's intentions :)

Reply:
currently this information is just for the autoscaler, but if i understand @sbueringer 's intention he is saying that in the future we might want to re-use these annotations for other things.

i think @vincepri 's suggestion is good for `capacity.cluster.x-k8s.io/<resource>`, we had discussed previously having the `machine` part in there as well to help indicate that is was per-machine. but if the community is happy with a short name, i think it makes the ux a little easier.

Reply:
> Note: I'm just guessing about @vincepri's intentions :). But for me it comes down to, either we want to make clear those fields are dedicated for the autoscaler or we open them up to other use cases. (I don't know what we want)

+1, i think the important information to convey is that these are resource capacity information about machines in the group. adding any mention of the "autoscaler" is helpful for understanding the usage, but is it something we want to bake into the API? (i can see the wisdom is dropping it, but we had requests previously to add it)

Reply:
no strong opinions, but if we are removing autoscaler from status, let's remove it from here as well

Reply:
> no strong opinions, but if we are removing autoscaler from status, let's remove it from here as well

+1. I also only think we should do it consistently

Reply:
agreed about keeping them consistent